syncer/syncer.go

```diff
@@ -1039,6 +1039,20 @@ func (s *Syncer) sync(tctx *tcontext.Context, queueBucket string, db *DBConn, jo
 			if !ok {
 				return
 			}
+
+			select {
+			case <-tctx.Ctx.Done():
+				// do not handle jobs anymore, because it should be failed with a done context.
+				// and avoid some errors like:
+				//  - `driver: bad connection` for `BEGIN`
+				//  - `sql: connection is already closed` for `BEGIN`
+				tctx.L().Debug("skip a remaining DML job in the job chan", zap.Stringer("job", sqlJob), zap.Int("remain", len(jobChan)))
+				s.execError.Set(tctx.Ctx.Err()) // set exexError to avoid flushing checkpoint because some jobs are skipped.
+				s.jobWg.Done()                  // done WaitGroup to avoid blocking
+				continue
```

Comment:
how about `break` here or use `fatalF`. `continue` could consume `jobChan` and return but code still might go to `case <-time.After(waitTime)` and execute already saved `jobs`

Reply:
move `tctx.Ctx.Done()` check into `executeSQLs` in https://github.com/pingcap/dm/pull/1295/commits/86a09c1b6e583a8e87c40355e6d24649aaeb8464.